src/main/scala/org/neo4j/spark/service/SchemaService.scala

```diff
@@ -331,20 +330,134 @@ class SchemaService(private val options: Neo4jOptions, private val driverCache:
     }
   }
 
-  def isReadQuery(query: String): Boolean = {
+  def isValidQuery(query: String, expectedQueryTypes: org.neo4j.driver.summary.QueryType*): Boolean = try {
     val queryType = session.run(s"EXPLAIN $query").consume().queryType()
-    queryType == org.neo4j.driver.summary.QueryType.READ_ONLY || queryType == org.neo4j.driver.summary.QueryType.SCHEMA_WRITE
+    expectedQueryTypes.size == 0 || expectedQueryTypes.contains(queryType)
+  } catch {
+    case e: Throwable => {
+      log.error("Query not compiled because of the following exception:", e)
+      false
+    }
   }
 
   def isValidQueryCount(query: String): Boolean = {
-    val resultSummary = session.run(s"EXPLAIN $query").consume()
-    val queryType = resultSummary.queryType()
-    val plan = resultSummary.plan()
-    val isReadOnly = queryType == org.neo4j.driver.summary.QueryType.READ_ONLY || queryType == org.neo4j.driver.summary.QueryType.SCHEMA_WRITE
-    val hasCountIdentifier = plan.identifiers().asScala.toSet == Set("count")
-    isReadOnly && hasCountIdentifier
+    try {
+      val resultSummary = session.run(s"EXPLAIN $query").consume()
+      val queryType = resultSummary.queryType()
+      val plan = resultSummary.plan()
+      val isReadOnly = queryType == org.neo4j.driver.summary.QueryType.READ_ONLY || queryType == org.neo4j.driver.summary.QueryType.SCHEMA_WRITE
+      val hasCountIdentifier = plan.identifiers().asScala.toSet == Set("count")
+      isReadOnly && hasCountIdentifier
+    } catch {
+      case e: Throwable => {
+        log.error("Query not compiled because of the following exception:", e)
+        false
+      }
+    }
   }
 
+  private def createIndexOrConstraint(action: OptimizationType.Value, label: String, props: Seq[String]) = {
+    if (action == OptimizationType.NONE) {
+      log.info("No optimization type provided")
+      Unit
+    } else {
```

Comment:
remove else block.   It makes code easier to return if up top you have the shortcut return nothing statement, then the rest of the method just flows without the else block.   Alternatively, you have a lot of other functions below that are just match statements on an input.  This one could be done the same, but the double nested try/catch/if/else blocks below make the code harder to read.

Reply:
done, but I did it in a "scala way" with pattern matching